airflow/dag_processing/manager.py

```diff
@@ -562,118 +568,144 @@ def _run_parsing_loop(self):
+                if span.is_recording():
+                    span.add_event(name="heartbeat()")
+                if self._direct_scheduler_conn is not None and self._direct_scheduler_conn in ready:
+                    agent_signal = self._direct_scheduler_conn.recv()
+
+                    self.log.debug("Received %s signal from DagFileProcessorAgent", agent_signal)
+                    if agent_signal == DagParsingSignal.TERMINATE_MANAGER:
+                        self.terminate()
+                        if span.is_recording():
+                            span.add_event(name="terminate()")
+                        break
+                    elif agent_signal == DagParsingSignal.END_MANAGER:
+                        self.end()
+                        if span.is_recording():
+                            span.add_event(name="end()")
+                        sys.exit(os.EX_OK)
+                    elif agent_signal == DagParsingSignal.AGENT_RUN_ONCE:
+                        # continue the loop to parse dags
+                        pass
+                    elif isinstance(agent_signal, CallbackRequest):
+                        self._add_callback_to_queue(agent_signal)
+                    else:
+                        raise ValueError(f"Invalid message {type(agent_signal)}")
+
+                if not ready and not self._async_mode:
+                    # In "sync" mode we don't want to parse the DAGs until we
+                    # are told to (as that would open another connection to the
+                    # SQLite DB which isn't a good practice
+
+                    # This shouldn't happen, as in sync mode poll should block for
+                    # ever. Lets be defensive about that.
+                    self.log.warning(
+                        "wait() unexpectedly returned nothing ready after infinite timeout (%r)!", poll_time
+                    )
 
-            self._print_stat()
+                    continue
 
-            all_files_processed = all(self.get_last_finish_time(x) is not None for x in self.file_paths)
-            max_runs_reached = self.max_runs_reached()
+                for sentinel in ready:
+                    if sentinel is not self._direct_scheduler_conn:
+                        processor = self.waitables.get(sentinel)
+                        if processor:
+                            self._collect_results_from_processor(processor)
+                            self.waitables.pop(sentinel)
+                            self._processors.pop(processor.file_path)
+
+                if self.standalone_dag_processor:
+                    self._fetch_callbacks(max_callbacks_per_loop)
+                self._scan_stale_dags()
+                DagWarning.purge_inactive_dag_warnings()
+                refreshed_dag_dir = self._refresh_dag_dir()
+
+                self._kill_timed_out_processors()
+                if span.is_recording():
+                    span.add_event(name="_kill_timed_out_processors()")
```

Comment:
This pattern is repeated a LOT.   For future consideration, I wonder if there might be a way to add a decorator to the classes which adds an event when it is called.  It would mean storing span as a class object (as self.span) and maybe something roughly along the lines of 

```
def addBasicTraceEvent(self, func):
    def wrapper(self, func):
        if self.span.is_recording():
            span.add_event(name=func.__name__)
    return wrapper
```    

then just adding the @addBasicTraceEvent decorator to each of these methods that don't have an attributes value.  Maybe that could be a future project for someone.

Reply:
> This pattern is repeated a LOT. For future consideration, I wonder if there might be a way to add a decorator to the classes which adds an event when it is called. It would mean storing span as a class object (as self.span) and maybe something roughly along the lines of
> 
> ```
> def addBasicTraceEvent(self, func):
>     def wrapper(self, func):
>         if self.span.is_recording():
>             span.add_event(name=func.__name__)
>     return wrapper
> ```
> 
> then just adding the @addBasicTraceEvent decorator to each of these methods that don't have an attributes value. Maybe that could be a future project for someone.

Yes, definitely think the pattern can be wrapped to a decorator. I'd like to defer this to the next project :-)